protocol/app/ve/applier/price_writer.go

```diff
@@ -105,112 +112,266 @@ func (pa *PriceApplier) getPricesAndAggregateFromVE(
 		}
+	}
 
-		if err := pa.pricesKeeper.UpdateMarketPrice(ctx, &newPrice); err != nil {
-			pa.logger.Error(
-				"failed to set price for currency pair",
-				"currency_pair", pair,
-				"err", err,
-			)
+	pa.finalPriceCache.SetPriceUpdates(ctx, finalPriceUpdates, round)
 
-			return err
-		}
+	return nil
+}
 
-		finalPriceUpdates.MarketPriceUpdates = append(finalPriceUpdates.MarketPriceUpdates, &newPrice)
+func (pa *PriceApplier) writePnlAndSpotPriceToStore(
+	ctx sdk.Context,
+	pricePair voteweighted.AggregatorPricePair,
+	marketId uint32,
+) (pricecache.PriceUpdate, error) {
+	newPrice := &pricestypes.MarketPriceUpdate{
+		MarketId:  marketId,
+		SpotPrice: pricePair.SpotPrice.Uint64(),
+		PnlPrice:  pricePair.PnlPrice.Uint64(),
+	}
 
-		pa.logger.Info(
-			"set price for currency pair",
-			"currency_pair", pair,
-			"quote_price", newPrice.Price,
+	if err := pa.pricesKeeper.UpdateSpotAndPnlMarketPrices(ctx, newPrice); err != nil {
+		pa.logger.Error(
+			"failed to set price for currency pair",
+			"market_id", marketId,
+			"err", err,
 		)
+
+		return pricecache.PriceUpdate{}, err
 	}
 
-	pa.finalPriceCache.SetPriceUpdates(ctx, finalPriceUpdates, round)
+	pa.logger.Info(
+		"set price for currency pair",
+		"market_id", marketId,
+		"spot_price", newPrice.SpotPrice,
+		"pnl_price", newPrice.PnlPrice,
+	)
+
+	finalPriceUpdate := pricecache.PriceUpdate{
+		MarketId:  marketId,
+		SpotPrice: pricePair.SpotPrice,
+		PnlPrice:  pricePair.PnlPrice,
+	}
 
-	return nil
+	return finalPriceUpdate, nil
 }
 
-func (pa *PriceApplier) shouldWritePriceToStore(
+func (pa *PriceApplier) writePnlPriceToStore(
 	ctx sdk.Context,
-	price *big.Int,
+	pricePair voteweighted.AggregatorPricePair,
 	marketId uint32,
-) (bool, *big.Int) {
+) (pricecache.PriceUpdate, error) {
+	pnlPriceUpdate := &pricestypes.MarketPnlPriceUpdate{
+		MarketId: marketId,
+		PnlPrice: pricePair.PnlPrice.Uint64(),
+	}
+
+	if err := pa.pricesKeeper.UpdatePnlPrice(ctx, pnlPriceUpdate); err != nil {
+		return pricecache.PriceUpdate{}, err
+	}
 
-	if price.Sign() == -1 {
+	pa.logger.Info(
+		"set price for currency pair",
+		"market_id", marketId,
+		"pnl_price", pricePair.PnlPrice.Uint64(),
+	)
+
+	return pricecache.PriceUpdate{
+		MarketId:  marketId,
+		SpotPrice: nil,
+		PnlPrice:  pricePair.PnlPrice,
+	}, nil
+}
+
+func (pa *PriceApplier) writeSpotPriceToStore(
+	ctx sdk.Context,
+	pricePair voteweighted.AggregatorPricePair,
+	marketId uint32,
+) (pricecache.PriceUpdate, error) {
+	spotPriceUpdate := &pricestypes.MarketSpotPriceUpdate{
+		MarketId:  marketId,
+		SpotPrice: pricePair.SpotPrice.Uint64(),
+	}
+
+	if err := pa.pricesKeeper.UpdateSpotPrice(ctx, spotPriceUpdate); err != nil {
+		return pricecache.PriceUpdate{}, err
+	}
+
+	pa.logger.Info(
+		"set price for currency pair",
+		"market_id", marketId,
+		"spot_price", pricePair.SpotPrice.Uint64(),
+	)
+
+	return pricecache.PriceUpdate{
+		MarketId:  marketId,
+		SpotPrice: pricePair.SpotPrice,
+		PnlPrice:  nil,
+	}, nil
+}
+
```

Comment:
yeah I did this, to avoid a long function with if statements, because in prices keeper you have different functions to adjust just spot, pnl, or both (this was done to avoid unnecessary computation. Thing is these take in different types so the function would be pretty long. Had it originally like this but separated for clean purposes, curious to hear your opinion now that you know rationale